File: common/mrtjp/projectred/transmission/APIImpl_Transmission.java
package mrtjp.projectred.transmission;

import java.util.ArrayList;
import java.util.List;

import mrtjp.projectred.api.IBundledEmitter;
import mrtjp.projectred.api.IBundledTile;
import mrtjp.projectred.api.ISpecialLinkState;
import mrtjp.projectred.api.ITransmissionAPI;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;
import codechicken.lib.vec.BlockCoord;
import codechicken.lib.vec.Rotation;
import codechicken.multipart.TMultiPart;
import codechicken.multipart.TileMultipart;

public class APIImpl_Transmission implements ITransmissionAPI
{
    @Override
    public byte[] getBundledInput(World world, int x, int y, int z, int side)
    {
        BlockCoord pos = new BlockCoord(x, y, z).offset(side);
        TileEntity t = world.getBlockTileEntity(pos.x, pos.y, pos.z);
        if (t instanceof IBundledTile)
            return ((IBundledTile) t).getBundledSignal(side ^ 1);
        else if (t instanceof TileMultipart)
        {
            TileMultipart tmp = (TileMultipart) t;
            byte[] signal = null;
            for (int r = 0; r < 4; r++)
            {
                int pside = Rotation.rotateSide(side, r);
                TMultiPart p = tmp.partMap(pside);
                if (p instanceof IBundledEmitter)
                    signal = BundledCableCommons.raiseSignal(signal, ((IBundledEmitter) p).getBundledSignal(Rotation.rotationTo(pside, side ^ 1)));
            }

            TMultiPart p = tmp.partMap(6);
            if (p instanceof IBundledEmitter)
                signal = BundledCableCommons.raiseSignal(signal, ((IBundledEmitter) p).getBundledSignal(side ^ 1));

            return signal;
        }

        return null;
    }
}

File: src/main/scala/mrtjp/projectred/transmission/APIImpl_Transmission.scala
package mrtjp.projectred.transmission

import codechicken.lib.vec.Rotation
import codechicken.multipart.{BlockMultipart, TileMultipart}
import mrtjp.projectred.api.{IBundledEmitter, IBundledTile, IBundledTileInteraction, ITransmissionAPI}
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

class APIImpl_Transmission extends ITransmissionAPI
{
    override def registerBundledTileInteraction(interaction:IBundledTileInteraction)
    {
        APIImpl_Transmission.interactions :+= interaction
    }

    override def getBundledInput(world:World, pos:BlockPos, facing:EnumFacing):Array[Byte] =
    {
        val side = facing.ordinal
        world.getTileEntity(pos.offset(EnumFacing.values()(side))) match {
            case ibt:IBundledTile => ibt.getBundledSignal(side^1)

            case tmp:TileMultipart =>
                var signal:Array[Byte] = null
                def raise(ibe:IBundledEmitter, r:Int) {
                    signal = BundledCommons.raiseSignal(signal, ibe.getBundledSignal(r))
                }

                for (r <- 0 until 4) {
                    val pside = Rotation.rotateSide(side, r)
                    tmp.partMap(pside) match {
                        case ibe:IBundledEmitter => raise(ibe, Rotation.rotationTo(pside, side^1))
                        case _ =>
                    }
                }
                tmp.partMap(6) match {
                    case ibe:IBundledEmitter => raise(ibe, side^1)
                    case _ =>
                }
                signal

            case _ => null
        }
    }

    override def containsBundledCable(world:World, pos:BlockPos, side:EnumFacing) =
        BlockMultipart.getPart(world, pos, side.ordinal()) match {
            case be:IBundledCablePart => true
            case _ => false
        }

    override def containsFramedWire(world:World, pos:BlockPos) =
        BlockMultipart.getPart(world, pos, 6).isInstanceOf[FramedWirePart]

    override def getFramedWireConnectionMask(world:World, pos:BlockPos) =
        BlockMultipart.getPart(world, pos, 6) match {
            case f:FramedWirePart => f.clientConnMap
            case _ => -1
        }
}

object APIImpl_Transmission
{
    var interactions = Vector[IBundledTileInteraction]()

    def isValidInteractionFor(world:World, pos:BlockPos, side:EnumFacing) =
        interactions.exists(_.isValidInteractionFor(world, pos, side))

    def canConnectBundled(world:World, pos:BlockPos, side:EnumFacing):Boolean =
        interactions.find(_.isValidInteractionFor(world, pos, side)) match {
            case Some(e) => e.canConnectBundled(world, pos, side)
            case None => false
        }

    def getBundledSignal(world:World, pos:BlockPos, side:EnumFacing):Array[Byte] =
        interactions.find(_.isValidInteractionFor(world, pos, side)) match {
            case Some(e) => e.getBundledSignal(world, pos, side)
            case None => null
        }
}
